src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs

```diff
@@ -71,19 +89,194 @@ protected override void EndProcessing()
             {
                 return;
             }
-            orderByProperty.OrderMatrix.Sort(orderByProperty.Comparer);
 
-            if (_unique)
+            // If -Unique was used, if -Top & -Bottom were not used, or of -Top or -Bottom would return all objects, sort
+            // all objects, remove duplicates if necessary, and identify the list of items to return
+            if (_unique || (Top == 0 && Bottom == 0) || Top >= orderByProperty.OrderMatrix.Count || Bottom >= orderByProperty.OrderMatrix.Count)
             {
-                RemoveDuplicates(orderByProperty);
+                orderByProperty.OrderMatrix.Sort(orderByProperty.Comparer);
+
+                if (_unique)
+                {
+                    RemoveDuplicates(orderByProperty);
+                }
+
+                var listToOutput = orderByProperty.OrderMatrix;
+
+                if (Top > 0 && Top < listToOutput.Count)
+                {
+                    // There may be a faster algorithm to sort while eliminating duplicates and retaining only
+                    // the top N records; for now though, this gets the job done
+                    listToOutput = listToOutput.GetRange(0, Top);
+                }
+                else if (Bottom > 0 && Bottom < listToOutput.Count)
+                {
+                    int lastItemIndex = listToOutput.Count - 1;
+                    listToOutput = listToOutput.GetRange(lastItemIndex - Bottom, lastItemIndex);
+                }
+
+                // write to output stream
+                foreach (var x in listToOutput)
+                {
+                    WriteObject(x.inputObject);
+                }
+            }
+            // Otherwise, use an indexed min-/max-heap to sort all objects
+            else
+            {
+                // The heap entry comparer will sort the items in the heap
+                var comparer = new IndexedMinMaxHeapComparer(orderByProperty.Comparer);
+
+                // Identify how many items will be in the heap
+                int indexedMinMaxHeapSize = Top > 0 ? Top : Bottom;
+
+                // A small list will store the data for the heap 
+                var indexedMinMaxHeap = new List<IndexedMinMaxHeapEntry>(indexedMinMaxHeapSize + 1);
```

Comment:
@iSazonov Every part of what is being done is necessary (although the code is changing/improving based on discussion with @lzybkr).
1. I'm not replacing standard sort code. The standard sort still exists. But for Top N/Bottom N sort, a heap-sort is necessary for performance reasons.
1. All sort results from Top N/Bottom N are being compared with the standard sort so that results are predictable, not random.
1. You're absolutely incorrect thinking that everyone knows that the .NET standard is a unstable sort. PowerShell's main audience are IT Pros, not developers.
1. Part of my goal with this PR is to allow someone to replace Sort-Object | Select-Object -First/-Last without any other code changes. Another part of my goal with this PR is to allow direct comparison between standard sort and Top N/Bottom N sort. Neither of those are possible without a stable sort.
I'll continue moving forward with the changes that I have been discussing with @lzybkr.